packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,285 @@
+import "../../common/implementation/Testable.sol";
+
+contract MerkleDistributor is Ownable, Lockable, Testable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    // A Window maps a Merkle root to a reward token address and a claim window.
+    // Claims for this window cannot take place before start time.
+    struct Window {
+        // Claims for this window can begin after `start`.
+        uint256 start;
+        // Claims become fully vested after `end`.
+        uint256 end;
+        // Merkle root describing the distribution.
+        bytes32 merkleRoot;
+        // Currency in which reward is processed.
+        IERC20 rewardToken;
+        // Total amount of rewards distributed this window. This is not enforced
+        // but might be useful to query.
+        uint256 totalRewardsDistributed;
+        // Owner can set this to true to block claims for this window.
+        bool locked;
+    }
+
+    // Represents an account's claim for `amount` within the Merkle root located at the `windowIndex`.
+    struct Claim {
+        uint256 windowIndex;
+        uint256 amount;
+        address account;
+        bytes32[] merkleProof;
+    }
+
+    // Windows are mapped to arbitrary indices.
+    mapping(uint256 => Window) public merkleWindows;
+
+    // Keep track of the amount an account has claimed from a window.
+    mapping(uint256 => mapping(address => uint256)) public amountClaimedFromWindow;
+
+    // Index of last seeded root. Next allocation to begin at `lastSeededIndex + 1`.
+    uint256 public lastSeededIndex;
+
+    // Events:
+    event Claimed(
+        address caller,
+        uint256 indexed windowIndex,
+        address indexed account,
+        uint256 totalClaimAmount,
+        uint256 amountVestedNetPreviousClaims,
+        uint256 claimAmountRemaining,
+        address indexed rewardToken
+    );
+    event SeededWindow(
+        uint256 indexed windowIndex,
+        uint256 amount,
+        uint256 indexed windowStart,
+        uint256 windowEnd,
+        address indexed rewardToken
+    );
+
+    modifier windowNotLocked(uint256 windowIndex) {
+        require(!merkleWindows[windowIndex].locked, "Window distributions locked");
+        _;
+    }
+
+    constructor(address _timerAddress) public Testable(_timerAddress) {}
+
+    /****************************
+     *
+     * Admin functions
+     *
+     ****************************/
+
+    // Set merkle root for the next available window index and seed allocations. Callable by owner of this
+    // contract. Importantly, we assume that the owner of this contract
+    // correctly chooses an amount `totalRewardsDistributed` that is sufficient
+    // to cover all claims within the `merkleRoot`. Otherwise, a race condition
+    // can be created either within a window (last claimants within a window don't get their claim)
+    // or across windows (claimants from window T take rewards from window T+1).
```

Comment:
How would a claimant approved for window T be able to claim extra rewards from window T+1? Wouldn't their claims for window T automatically expire when the window ends?

Reply:
Actually if the `current time > end timestamp T` then the full amount can be claimed. Rewards never "expire" currently, they become fully vested after the window ends.

This is the code from `calcVestedAmount`

```
        // If time is after window end or window end is not greater than start,then return all tokens.
        else if (time >= windowEnd || windowEnd <= windowStart) {
            return amount;

```

Reply:
Actually claimed rewards never expire and after a window ends, the amount just gets fully vested instead of retired. See this code from `calcVestedAmount`:

```
        // If time is after window end or window end is not greater than start,then return all tokens.
        else if (time >= windowEnd || windowEnd <= windowStart) {
            return amount;
```

Reply:
Gotcha, that makes sense. I'm still not sure how claimants from window T could take rewards from window T+1. Can you explain that one?

Reply:
This would arise from an undesired race condition. Imagine the following situation:

Window 1: Deposit 100 tokens and inject a Merkle Proof that gives claimant A 50 tokens and claimant B 51 tokens. (This is an error on the contract owner's part for not setting the `totalRewardAmount` of 100 correctly to 101)

Window 2: Deposit 1 token and inject a Merkle Proof that gives claimant A 1 token.

At this point contract has 100 +1 = 101 tokens.

Now, imagine the following sequence of claims:
- Claimant A claims 50 tokens for window 1 (contract now has 101 - 50 = 51 tokens)
- Claimant B claims 51 tokens for window 1 (contract now has 51 - 51 = 0 tokens)
- Claimant A now tries to claim 1 token for window 2 but fails

Thus, the contract Owner allowed "window 1 to steal from window 2"

How could I change the comments to better describe this?

Reply:
This is a really clear example. Maybe you could include a stripped-down version of the example as a comment?

Reply:
Done